cdc/sink/mq.go

```diff
@@ -261,5 +234,125 @@ func newKafkaSaramaSink(ctx context.Context, sinkURI *url.URL, filter *util.Filt
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return newMqSink(producer, filter, opts), nil
+	return newMqSink(producer, filter, replicaConfig, opts), nil
+}
+
+type partitionRule int
+
+const (
+	partitionRuleDefault partitionRule = iota
+	partitionRuleRowID
+	partitionRuleTS
+	partitionRuleTable
+)
+
+func (r *partitionRule) fromString(rule string) {
+	switch strings.ToLower(rule) {
+	case "default":
+		*r = partitionRuleDefault
+	case "rowid":
+		*r = partitionRuleRowID
+	case "ts":
+		*r = partitionRuleTS
+	case "table":
+		*r = partitionRuleTable
+	default:
+		*r = partitionRuleDefault
+		log.Warn("can't support partition rule, using default rule", zap.String("rule", rule))
+	}
+}
+
+type partitioner struct {
+	rules         map[entry.TableName]partitionRule
+	caseSensitive bool
+	partitionNum  int32
+}
+
+func newPartitioner(config *util.ReplicaConfig, partitionNum int32) *partitioner {
+	p := &partitioner{
+		caseSensitive: config.FilterCaseSensitive,
+		partitionNum:  partitionNum,
+		rules:         make(map[entry.TableName]partitionRule, len(config.SinkPartitionRules)),
+	}
+	for _, ruleConfig := range config.SinkPartitionRules {
+		tableName := entry.TableName{Schema: ruleConfig.Schema, Table: ruleConfig.Name}
+		if !p.caseSensitive {
+			tableName.Schema = strings.ToLower(tableName.Schema)
+			tableName.Table = strings.ToLower(tableName.Table)
+		}
+		var rule partitionRule
+		rule.fromString(ruleConfig.Rule)
+		p.rules[tableName] = rule
+	}
+	return p
+}
+
+func (p *partitioner) calPartition(row *model.RowChangedEvent) int32 {
+	rule, exist := p.rules[entry.TableName{Schema: row.Schema, Table: row.Table}]
+	if !exist {
+		rule = partitionRuleDefault
+	}
+	switch rule {
+	case partitionRuleDefault:
+		return p.defaultPartition(row)
+	case partitionRuleRowID:
+		return p.rowIDPartition(row)
+	case partitionRuleTS:
+		return p.tsPartition(row)
+	case partitionRuleTable:
+		return p.tablePartition(row)
+	default:
+		panic("unreachable")
+	}
+}
+
+func (p *partitioner) defaultPartition(row *model.RowChangedEvent) int32 {
+	hash := crc32.NewIEEE()
+	// distribute partition by table
+	_, err := hash.Write([]byte(row.Schema))
+	if err != nil {
+		log.Fatal("calculate hash of message key failed, please report a bug", zap.Error(err))
+	}
+	_, err = hash.Write([]byte(row.Table))
+	if err != nil {
+		log.Fatal("calculate hash of message key failed, please report a bug", zap.Error(err))
+	}
```

Comment:
Should the default partition strategy work as follows:
- If `IndieMarkCol` exists, use rowid or unique column value to calculate partition
- Otherwise, use schema.table to calculate partition.

`schema` and `table` are always used to calculate now

Reply:
It's fine